File: frontend/src/api/grn.ts
import api from './index';

// ============================================================================
// TYPES
// ============================================================================

export interface GRNItemUpdate {
    item_id: number;
    gross_received: number;
    damage: number;
    reject: number;
    damage_cost_allocation?: 'farm' | 'us';
    reject_cost_allocation?: 'farm' | 'us';
    notes?: string;
}

export interface GRNUpdateRequest {
    transport_method?: string;
    number_of_boxes?: number;
    receiving_time?: string;  // HH:MM:SS
    receiving_date?: string;   // YYYY-MM-DD
    receiver_id?: number;
    items?: GRNItemUpdate[];
    notes?: string;
}

export interface GRNItemResponse extends GRNItemUpdate {
    id: number;
    item_name: string;
    final_accepted: number;
    added_to_po: boolean;
    expected_quantity: number; // Added based on context in pages
}

export interface GRNPhotoResponse {
    id: number;
    item_id: number;
    item_name: string;
    photo_type: 'damage' | 'reject';
    photo_url: string;
    uploaded_by: number;
    uploaded_by_name: string;
    uploaded_at: string;
    file_size: number;
    notes?: string;
}

export interface GRNResponse {
    id: number;
    grn_number: string;
    po_id: number;
    po_number: string;
    batch_id: number;
    batch_number: string;
    vendor_id: number;
    vendor_name: string;
    transport_method?: string;
    number_of_boxes?: number;
    receiving_time?: string;
    receiving_date: string;
    status: string;
    receiver_id?: number;
    receiver_name?: string;
    created_at: string;
    completed_at?: string;
    notes?: string;
}

export interface EditHistory {
    field_name: string;
    old_value?: string;
    new_value: string;
    edited_by: number;
    edited_by_name: string;
    edited_at: string;
}

export interface GRNDetailResponse extends GRNResponse {
    items: GRNItemResponse[];
    photos: GRNPhotoResponse[];
    edit_history: EditHistory[];
}

export interface GRNListParams {
    status?: string;
    po_id?: number;
    batch_number?: string;
    from_date?: string;
    to_date?: string;
    page?: number;
    limit?: number;
}

export interface GRNListResponse {
    grns: GRNResponse[];
    total: number;
    page: number;
    limit: number;
}

// ============================================================================
// API FUNCTIONS
// ============================================================================

export const grnAPI = {
    // GRN Management
    generate: (poId: number) =>
        api.post<GRNResponse>('/grn/generate', { po_id: poId }),

    getById: (grnId: number) =>
        api.get<GRNDetailResponse>(`/grn/${grnId}`),

    update: (grnId: number, data: GRNUpdateRequest) =>
        api.put<GRNResponse>(`/grn/${grnId}/update`, data),

    finalize: (grnId: number) =>
        api.post<GRNResponse>(`/grn/${grnId}/finalize`),

    list: (params: GRNListParams) =>
        api.get<GRNListResponse>('/grn/list', { params }),

    getByPO: (poId: number) =>
        api.get<GRNResponse[]>(`/grn/by-po/${poId}`),

    getByBatch: (batchNumber: string) =>
        api.get<GRNDetailResponse>(`/grn/by-batch/${batchNumber}`),

    generateBlankPDF: (grnId: number) =>
        api.get(`/grn/${grnId}/print`, { responseType: 'blob' }),

    // Photo Management
    uploadPhotos: (grnId: number, itemId: number, photoType: 'damage' | 'reject', files: File[]) => {
        const formData = new FormData();
        files.forEach(file => formData.append('files', file));

        return api.post<string[]>(
            `/grn/${grnId}/photos/upload?item_id=${itemId}&photo_type=${photoType}`,
            formData,
            {
                headers: { 'Content-Type': 'multipart/form-data' }
            }
        );
    },

    deletePhoto: (photoId: number) =>
        api.delete(`/grn/photos/${photoId}`),
};

File: backend/app/schemas/grn.py

from pydantic import BaseModel, Field, validator
from typing import Optional, List
from datetime import date, time, datetime
from decimal import Decimal

class GRNItemUpdate(BaseModel):
    item_id: int
    gross_received: Decimal = Field(..., ge=0, decimal_places=3)
    damage: Decimal = Field(default=0, ge=0, decimal_places=3)
    reject: Decimal = Field(default=0, ge=0, decimal_places=3)
    damage_cost_allocation: Optional[str] = Field(None, pattern='^(farm|us)$')
    reject_cost_allocation: Optional[str] = Field(None, pattern='^(farm|us)$')
    notes: Optional[str] = None
    
    @validator('damage_cost_allocation')
    def validate_damage_allocation(cls, v, values):
        if values.get('damage', 0) > 0 and not v:
            raise ValueError('Cost allocation required when damage > 0')
        return v
    
    @validator('reject_cost_allocation')
    def validate_reject_allocation(cls, v, values):
        if values.get('reject', 0) > 0 and not v:
            raise ValueError('Cost allocation required when reject > 0')
        return v

class GRNUpdateRequest(BaseModel):
    transport_method: Optional[str] = None
    number_of_boxes: Optional[int] = Field(None, ge=0)
    receiving_time: Optional[time] = None
    receiving_date: Optional[date] = None
    receiver_id: Optional[str] = None
    items: Optional[List[GRNItemUpdate]] = None
    notes: Optional[str] = None

class GRNItemResponse(BaseModel):
    id: int
    item_id: int
    item_name: Optional[str] = None # fetched from join
    gross_received: Decimal
    damage: Decimal
    reject: Decimal
    final_accepted: Decimal
    damage_cost_allocation: Optional[str] = None
    reject_cost_allocation: Optional[str] = None
    notes: Optional[str] = None
    added_to_po: bool

class GRNPhotoResponse(BaseModel):
    id: int
    item_id: Optional[int] = None
    item_name: Optional[str] = None
    photo_type: str
    photo_url: str
    uploaded_by: str
    uploaded_by_name: Optional[str] = None
    uploaded_at: datetime
    file_size: Optional[int] = None
    notes: Optional[str] = None

class GRNResponse(BaseModel):
    id: int
    grn_number: str
    po_id: int
    po_number: Optional[str] = None
    batch_id: int
    batch_number: Optional[str] = None
    vendor_id: Optional[int] = None # derived from PO
    vendor_name: Optional[str] = None
    transport_method: Optional[str] = None
    number_of_boxes: Optional[int] = None
    receiving_time: Optional[time] = None
    receiving_date: date
    status: str
    receiver_id: Optional[str] = None
    receiver_name: Optional[str] = None
    created_at: datetime
    completed_at: Optional[datetime] = None
    notes: Optional[str] = None

class EditHistory(BaseModel):
    field_name: str
    old_value: Optional[str] = None
    new_value: Optional[str] = None
    edited_by: str
    edited_by_name: Optional[str] = None
    edited_at: datetime

class GRNDetailResponse(GRNResponse):
    items: List[GRNItemResponse]
    photos: List[GRNPhotoResponse]
    edit_history: List[EditHistory]

class GRNListResponse(BaseModel):
    grns: List[GRNResponse]
    total: int
    page: int
    limit: int
